### PR TITLE
Shared model crypto interfaces:

### PR DESCRIPTION
**Pull request by luckychess**
## What is this pull request?
Crypto interfaces for shared model.
   
## Why do you implement it? Why do we need this pull request?
At the moment shared model uses only one interface for all crypto related stuff - hashes, signatures, keys etc. This makes possible to mix data types and use private key instead of public key and so on by a mistake and it's impossible to detect on a compilation phase. Now we have dedicate types for all crypto primitives.
  
## Details/Features
- Rename old Hash interface for Blob
- Add interfaces for hash, public key, private key, signed data and keypair
- Use introduced interfaces in other shared model interfaces

### CODE REVIEW

File: shared_model/cryptography/blob.hpp [hunk 1]
```diff
@@ -15,37 +15,37 @@
...
  */
 
-#ifndef IROHA_SHARED_MODEL_HASH_HPP
-#define IROHA_SHARED_MODEL_HASH_HPP
+#ifndef IROHA_SHARED_MODEL_BLOB_HPP
+#define IROHA_SHARED_MODEL_BLOB_HPP
 
 #include <string>
 #include "interfaces/primitive.hpp"
 #include "model/signature.hpp"
```
Comment (l4l) [inline shared_model/cryptography/blob.hpp]: Don't need that line

File: shared_model/cryptography/keypair.hpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IROHA_SHARED_MODEL_KEYPAIR_HPP
+#define IROHA_SHARED_MODEL_KEYPAIR_HPP
+
+#include "interfaces/primitive.hpp"
+#include "cryptography/public_key.hpp
```
Comment (l4l) [inline shared_model/cryptography/keypair.hpp]: missing commas

File: shared_model/cryptography/keypair.hpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+namespace shared_model {
+  namespace crypto {
+    /**
+     * Class for holding a keypair: public key and private key
+     */
+    class Keypair : public Primitive<Keypair, Keypair> {
+     public:
+      /// Type of public key
+      using PublicKeyType = PublicKey;
```
Comment (l4l) [inline shared_model/cryptography/keypair.hpp]: Why not use just `PublicKey` type?

File: shared_model/cryptography/keypair.hpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      using PrivateKeyType = PrivateKey;
+
+      /**
+       * @return private key
+       */
+      virtual const PrivateKeyType &privateKey() const = 0;
+
+      bool operator==(const Keypair &keypair) const override {
+        return this->publicKey() == keypair.publicKey()
+            and this->privateKey() == keypair.privateKey();
```
Comment (l4l) [inline shared_model/cryptography/keypair.hpp]: remove `this->` here and above

File: shared_model/cryptography/blob.hpp [hunk 5]
```diff
@@ -15,37 +15,35 @@
...
 
 namespace shared_model {
-  namespace interface {
+  namespace crypto {
 
     /**
-     * Hash interface present user-friendly blob for working with low-level
-     * stuff, such as cryptography and object hashing
+     * Blob interface present user-friendly blob for working with low-level
+     * binary stuff
```
Comment (muratovv) [inline shared_model/cryptography/blob.hpp]: Also, there is need to write about not-fixing length. I mean in compile-time.

File: shared_model/cryptography/blob.hpp [hunk 6]
```diff
@@ -15,37 +15,35 @@
...
      */
-    class Hash : public Primitive<Hash, Hash> {
+    class Blob : public interface::Primitive<Blob, Blob> {
```
Comment (muratovv) [inline shared_model/cryptography/blob.hpp]: Also, you may improve interface with`toString`, `operator==` methods.

File: shared_model/cryptography/hash.hpp [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+#ifndef IROHA_SHARED_MODEL_HASH_HPP
+#define IROHA_SHARED_MODEL_HASH_HPP
+
+#include "cryptography/blob.hpp"
+#include "utils/string_builder.hpp"
+
+namespace shared_model {
+  namespace crypto {
+    /**
+     * A special class for storing hashes. Does not contain any special stuff.
```
Comment (muratovv) [inline shared_model/cryptography/hash.hpp]: >Does not contain any special stuff.
I am confused about explanation, maybe we can improve it?

File: shared_model/cryptography/keypair.hpp [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+#include "cryptography/public_key.hpp"
+#include "interfaces/primitive.hpp"
+#include "utils/string_builder.hpp"
+
+namespace shared_model {
+  namespace crypto {
+    /**
+     * Class for holding a keypair: public key and private key
+     */
+    class Keypair : public interface::Primitive<Keypair, Keypair> {
```
Comment (muratovv) [inline shared_model/cryptography/keypair.hpp]: Add todo about fix `Primitive` design.

File: shared_model/cryptography/keypair.hpp [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+     public:
+      /**
+       * @return public key
+       */
+      virtual const PublicKey &publicKey() const = 0;
```
Comment (muratovv) [inline shared_model/cryptography/keypair.hpp]: Add using for `PublicKey`.

File: shared_model/cryptography/keypair.hpp [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+      /**
+       * @return private key
+       */
+      virtual const PrivateKey &privateKey() const = 0;
```
Comment (muratovv) [inline shared_model/cryptography/keypair.hpp]: Add using for `PrivateKey`.

File: shared_model/cryptography/private_key.hpp [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+#ifndef IROHA_SHARED_MODEL_PRIVATE_KEY_HPP
+#define IROHA_SHARED_MODEL_PRIVATE_KEY_HPP
+
+#include "cryptography/blob.hpp"
+#include "utils/string_builder.hpp"
+
+namespace shared_model {
+  namespace crypto {
+    /**
+     * A special class for storing private keys. Does not contain any special
```
Comment (muratovv) [inline shared_model/cryptography/private_key.hpp]: >Does not contain any special stuff here.
Also, comment looks confusing.

File: shared_model/cryptography/private_key.hpp [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+     * stuff here.
+     */
+    class PrivateKey : public Blob {
+     public:
+      std::string toString() const override {
+        return detail::PrettyStringBuilder()
+            .init("PrivateKey")
+            .append(Blob::hex())
```
Comment (muratovv) [inline shared_model/cryptography/private_key.hpp]: Rework with constant string

File: shared_model/cryptography/public_key.hpp [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+#ifndef IROHA_SHARED_MODEL_PUBLIC_KEY_HPP
+#define IROHA_SHARED_MODEL_PUBLIC_KEY_HPP
+
+#include "cryptography/blob.hpp"
+#include "utils/string_builder.hpp"
+
+namespace shared_model {
+  namespace crypto {
+    /**
+     * A special class for storing public keys. Does not contain any special
```
Comment (muratovv) [inline shared_model/cryptography/public_key.hpp]: > Does not contain any special stuff here.
Also, confusing.

File: shared_model/cryptography/signed.hpp [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+
+#ifndef IROHA_SHARED_MODEL_SIGNED_HPP
+#define IROHA_SHARED_MODEL_SIGNED_HPP
+
+#include "cryptography/blob.hpp"
+#include "utils/string_builder.hpp"
+
+namespace shared_model {
+  namespace crypto {
+    class Signed : public Blob {
```
Comment (muratovv) [inline shared_model/cryptography/signed.hpp]: Add doc for class.

File: shared_model/interfaces/common_objects/signature.hpp [hunk 15]
```diff
@@ -32,19 +34,24 @@ namespace shared_model {
...
       /**
        * @return public key of signatory
        */
-      virtual const HashType &publicKey() const = 0;
+      virtual const PublicKeyType &publicKey() const = 0;
+
+      /**
+       * Type of signed hash
+       */
+      using SignedHashType = crypto::Signed;
```
Comment (muratovv) [inline shared_model/interfaces/common_objects/signature.hpp]: Rework name, pls.

File: shared_model/cryptography/keypair.hpp (hunk 3 above)
Comment (muratovv) [reply shared_model/cryptography/keypair.hpp]: For easy using in `auto` expressions.

**Pull request merged**